### ISSUE TITLE
local model

### ISSUE THREAD

**Issue by bipinpeter** [zzbuzzard/stable-diffusion-infinite-scroll#1]
how do i call a local SD model?
i made a folder 'model' placed a sd model, 
gave this command - python sd_scroll.py "fantasy mushroom forest|peaceful landscape" -s 40 -d H -m model/dreamshaper_8.safetensors

but gets this error
Error: Cannot load model model/dreamshaper_8.safetensors: model is not cached locally and an error occured while trying to fetch metadata from the Hub. Please check out the root cause in the stacktrace above

Loading SD...
Couldn't connect to the Hub: 401 Client Error. (Request ID: Root=1-65b35d28-760ad80923d49e461963a291;8a2db248-7275-4c25-9b0b-4eeb6d17501a)

Repository Not Found for url: https://huggingface.co/api/models/model/dreamshaper_8.safetensors/revision/fp16.
Please make sure you specified the correct `repo_id` and `repo_type`.

**Comment by zzbuzzard** [comment]
Hey, thanks for letting me know! I've added support for local models now (at least it works on my machine), so please update the repo and try again.

Let me know how it goes! Most non-inpainting SD models perform really badly at inpainting, so unfortunately if you do get it working there's a chance it doesn't work very well anyway (there might be clear boundary lines between images) - not sure if there's anything I can do about this.

P.S. if you get an error about number of channels, try passing  `--legacy` at the command line.

**Comment by bipinpeter** [comment]
thank you! will update and test.. will let you know

**Comment by HamletEagle** [comment]
@zzbuzzard is it possible to use SDXL inpainting checkpoints like this one https://civitai.com/models/403361/juggernaut-xl-inpainting?dialog=commentThread&commentId=396827?

I tried to use it and got an error:
`Loading SD...

tokenizer_config.json: 100%|██████████████████████████████████████████████████████████████████| 905/905 [00:00<?, ?B/s]
vocab.json: 100%|███████████████████████████████████████████████████████████████████| 961k/961k [00:00<00:00, 1.87MB/s]
merges.txt: 100%|███████████████████████████████████████████████████████████████████| 525k/525k [00:00<00:00, 1.38MB/s]
special_tokens_map.json: 100%|█████████████████████████████████████████████████████████| 389/389 [00:00<00:00, 393kB/s]
tokenizer.json: 100%|█████████████████████████████████████████████████████████████| 2.22M/2.22M [00:00<00:00, 4.36MB/s]
config.json: 100%|████████████████████████████████████████████████████████████████████████| 4.52k/4.52k [00:00<?, ?B/s]
Some weights of the model checkpoint were not used when initializing CLIPTextModel:
 ['text_model.embeddings.position_ids']
tokenizer_config.json: 100%|██████████████████████████████████████████████████████████████████| 904/904 [00:00<?, ?B/s]
vocab.json: 100%|███████████████████████████████████████████████████████████████████| 862k/862k [00:00<00:00, 50.5MB/s]
merges.txt: 100%|███████████████████████████████████████████████████████████████████| 525k/525k [00:00<00:00, 1.24MB/s]
special_tokens_map.json: 100%|████████████████████████████████████████████████████████████████| 389/389 [00:00<?, ?B/s]
tokenizer.json: 100%|█████████████████████████████████████████████████████████████| 2.22M/2.22M [00:00<00:00, 16.0MB/s]
config.json: 100%|████████████████████████████████████████████████████████████████| 4.88k/4.88k [00:00<00:00, 4.92MB/s]
Some weights of the model checkpoint were not used when initializing CLIPTextModelWithProjection:
 ['text_model.embeddings.position_ids']
Process Process-1:
Traceback (most recent call last):
  File "Python\Python310\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "Python\Python310\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "D:\stable-diffusion-infinite-scroll\sd_scroll.py", line 50, in generate_loop
    pipe = util.get_pipe(model, attn_slicing, legacy)
  File "D:\stable-diffusion-infinite-scroll\util.py", line 92, in get_pipe
    pipe = loader.from_single_file(
  File "D:\infinite_scroll\lib\site-packages\huggingface_hub\utils\_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
  File "D:\infinite_scroll\lib\site-packages\diffusers\loaders\single_file.py", line 313, in from_single_file
    pipe = pipeline_class(**init_kwargs)
TypeError: StableDiffusionInpaintPipeline.__init__() got an unexpected keyword argument 'tokenizer_2'`

--legacy did not help.

Thanks!

**Comment by zzbuzzard** [comment]
Hey @HamletEagle I've tried to add support for XL models just now, but unfortunately I'm unable to test it locally as my laptop doesn't have enough RAM or GPU VRAM for XL 😅 would you be able to check whether it works with the new code? Have added a new flag `--xl`, so the full command would be something like

`python sd_scroll.py "prompt1|prompt2" -s 20 -r 768 --xl -m [path to downloaded XL safetensors file]`

Let me know!

**Comment by HamletEagle** [comment]
There are still some issues, but there is definitely progress:
```
Starting draw
Process Process-1:
Gap on right / bottom
Traceback (most recent call last):
  File "Python\Python310\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "Python\Python310\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "stable-diffusion-infinite-scroll\sd_scroll.py", line 65, in generate_loop
    front = next_image(pipe, image=front, base_size=base_size, prompt=prompt, shiftx=shiftx, shifty=shifty,
  File "stable-diffusion-infinite-scroll\util.py", line 32, in next_image
    assert image.size == (base_size, base_size)
AssertionError
```

Edit: using -r 1024 instead of -r 768 got it to work. It seems that image.size is (1024, 1024), so base_size can't be 768. Not sure why.

**Comment by zzbuzzard** [comment]
Thanks for testing this. Had a look and seems like the XL pipeline somehow fails to deduce the image size from the mask and defaults to 1024 x 1024 (as you noticed).

Have updated to tell it explicitly the size, hopefully it'll work now for any resolution...?

**Comment by HamletEagle** [comment]
Thank you for the quick updates and support. 

I tested the new code, and I have the following warnings, I am not sure if they are important, but maybe it shows an issue with correctly loading the model.
```
Some weights of the model checkpoint were not used when initializing CLIPTextModel:
 ['text_model.embeddings.position_ids']
Some weights of the model checkpoint were not used when initializing CLIPTextModelWithProjection:
 ['text_model.embeddings.position_ids']
```

The code runs successfully, but the quality is quite bad with settings that would produce high-quality results with this model. Perhaps the warning above is relevant and impacting the quality? There were 6 prompts used here with n = 6, r = 768, sh = 384

![out](https://github.com/zzbuzzard/stable-diffusion-infinite-scroll/assets/8179919/2d990bae-ec0f-4c73-870f-d01ecc07e4c6)


For example, I am running it with these 6 prompts generated by Claude3 as an example:
```
The Count of Monte Cristo, standing tall and victorious, satisfied smile, his enemies defeated, revenge complete, dramatic lighting emphasizing his triumphant pose, textures of rich fabrics and polished leather., hyper-realistic, 4k, ultra quality, cinematic, photorealistic
  The opulent interior of a grand Parisian ballroom, with the Count of Monte Cristo at the center. Ornate chandeliers cast a warm glow, elegantly dressed guests, expressions of admiration and envy., hyper-realistic, 4k, ultra quality, cinematic, photorealistic
  The Count's lavish estate, with the ballroom visible through grand windows. The estate grounds are meticulously landscaped, featuring perfectly manicured gardens, fountains, and sculptures, all bathed in the golden light of sunset., hyper-realistic, 4k, ultra quality, cinematic, photorealistic
  The Parisian cityscape, with the Count's estate nestled among the iconic architecture of the city. The Eiffel Tower stands tall in the distance, and the Seine River winds through the city, reflecting the warm hues of the sky., hyper-realistic, 4k, ultra quality, cinematic, photorealistic
  The entirety of Paris, with the Count's estate now a tiny speck amidst the sprawling city. The urban landscape is dotted with famous landmarks, such as the Louvre, Notre-Dame Cathedral, and the Arc de Triomphe, all illuminated by the fading light of dusk, hyper-realistic, 4k, ultra quality, cinematic, photorealistic
  The French countryside. Rolling hills, lush vineyards, and picturesque villages stretch out to the horizon, creating a sense of the Count's far-reaching influence and the scale of his triumphant journey., hyper-realistic, 4k, ultra quality, cinematic, photorealistic
```

There is a similar extension for infinite zooming (https://github.com/v8hid/infinite-zoom-automatic1111-webui) and with the above prompts, I was able to get something that looked quite nice. I hoped that I could use your code as a pair for infinite zoom, but instead of zooming, we would pan the image horizontally (I can do the panning part in a video editor after I get a good long image).


https://github.com/zzbuzzard/stable-diffusion-infinite-scroll/assets/8179919/a963788c-4dde-4878-8be4-06b81fcb9aa4

**Comment by zzbuzzard** [comment]
@HamletEagle Ah damn that's annoying, sorry about that. I suspect what's going on here is I'm just missing features - a nice thing about writing a A1111 extension is you get all the bells and whistles included, but here I have a pretty simple setup, for example it always uses the DDIM scheduler (in my defence, it was a good option when I wrote the original version of this years ago!). That will be easy to change as a command line option, I will do that at some point, but the issue here could be a lot of things really.

I tried to test using this model in google colab but even that didn't have enough RAM haha. Unfortunately I got a last deadline for my degree at the end of May so I won't have any time til then, but might look into this after that - maybe improving the quality here could improve it for other models too which would be nice.

P.S. Thanks for sharing the zoom video, v cool! Clearly a bit of a difference in quality lol

(Edit: btw I'm not certain but I believe those warnings are not the issue - although I'm a bit surprised that those particular weights are missing, this is caused by the checkpoint itself and as far as I understand it just means not every part of the CLIP text encoder needs to be used in this context, so not all weights are included)
